pkg/controller/apimanagerbackup/apimanagerbackup_logic_reconciler.go

```diff
@@ -0,0 +1,398 @@
+		options, err := apiManagerBackupOptionsProvider.Options()
+		if err != nil {
+			return nil, err
+		}
+		apiManagerBackup := backup.NewAPIManagerBackup(options)
+
+		res.apiManagerBackup = apiManagerBackup
+	}
+
+	return res, nil
+}
+
+func (r *APIManagerBackupLogicReconciler) Logger() logr.Logger {
+	return r.logger
+}
+
+func (r *APIManagerBackupLogicReconciler) Reconcile() (reconcile.Result, error) {
+	if r.cr.BackupCompleted() {
+		r.Logger().Info("Backup completed. End of reconciliation")
+		return reconcile.Result{}, nil
+	}
+
+	result, err := r.reconcileAPIManagerSourceStatusField()
+	if result.Requeue || err != nil {
+		return result, err
+	}
+
+	result, err = r.reconcileStartTimeField()
+	if result.Requeue || err != nil {
+		return result, err
+	}
+
+	result, err = r.reconcileBackupInS3Destination()
+	if result.Requeue || err != nil {
+		return result, err
+	}
+
+	result, err = r.reconcileBackupInPVCDestination()
+	if result.Requeue || err != nil {
+		return result, err
+	}
+
+	result, err = r.reconcileBackupCompletion()
+	if result.Requeue || err != nil {
+		return result, err
+	}
+
+	return result, err
+}
+
+func (r *APIManagerBackupLogicReconciler) reconcileBackupInS3Destination() (reconcile.Result, error) {
+	// TODO implement
+	return reconcile.Result{}, nil
+}
+
+func (r *APIManagerBackupLogicReconciler) reconcileBackupInPVCDestination() (reconcile.Result, error) {
+	var res reconcile.Result
+	var err error
+
+	err = r.reconcileBackupDestinationPVC()
+	if err != nil {
+		return reconcile.Result{}, err
+	}
+
+	res, err = r.reconcileBackupDestinationPVCStatus()
+	if res.Requeue || err != nil {
+		return res, err
+	}
+
+	res, err = r.reconcileBackupSecretsAndConfigMapsToPVCJob()
+	if res.Requeue || err != nil {
+		return res, err
+	}
+
+	res, err = r.reconcileAPIManagerCustomResourceBackupToPVCJob()
+	if res.Requeue || err != nil {
+		return res, err
+	}
+
+	res, err = r.reconcileBackupSystemFileStoragePVCToPVCJob()
+	if res.Requeue || err != nil {
+		return res, err
+	}
+
+	return res, err
+}
+
+func (r *APIManagerBackupLogicReconciler) reconcileBackupDestinationPVC() error {
+	desired := r.apiManagerBackup.BackupDestinationPVC()
+	if desired == nil {
+		return nil
+	}
+
+	// TODO create mutator function for PVC ?
+	err := r.ReconcileResource(&v1.PersistentVolumeClaim{}, desired, reconcilers.CreateOnlyMutator)
+
+	return err
+}
+
+func (r *APIManagerBackupLogicReconciler) reconcileBackupSecretsAndConfigMapsToPVCJob() (reconcile.Result, error) {
+	desired := r.apiManagerBackup.BackupSecretsAndConfigMapsToPVCJob()
+	if desired == nil {
+		return reconcile.Result{}, nil
+	}
+
+	// TODO create Mutator function for Jobs??
+	existing := &batchv1.Job{}
+	// Check if this step has finished
+	if r.cr.SecretsAndConfigMapsBackupStepFinished() {
+		return reconcile.Result{}, nil
+	}
+
+	// Check if backup substep has completed
+	if !r.cr.SecretsAndConfigMapsBackupSubStepFinished() {
```

Comment:
`r.cr.SecretsAndConfigMapsBackupStepFinished()`

`r.cr.SecretsAndConfigMapsBackupSubStepFinished()`

names are very confusing. What is that substep about? Maybe a better name?

Reply:
I agree with you that names are very confusing. That's the inherent problem of placing them in the same level of structure or in conditions where you have to namespace the names themselves and you cannot really define structure for them.
I could not come up with more clear names having them all in the same level.

An alternative I thought was to just create structure for them as I commented you in the https://github.com/3scale/3scale-operator/pull/358#discussion_r421516552 discussion. Allowing us to namespace the state and give it structure and potentially having other contextual information for some of them. Additionally, we could end up wanting more complex information than just a state, that's why I was thinking whether it would be a good idea to structure more the status, like:

.status.pvcDestination.configmapsAndSecrets.BackupFinished and other kind of potential information like .status.pvcDestination.logs

Of course this means that we would incorporate it as part of the CRD schema (although that's done also with conditions or fields at the same level  I think), but maybe being all of them optional would be acceptable? Of course in that case too we would have to perform nil checks in several places around the code as with any optional nested fields.

What do you think? As you suggested in another comment I totally think we should do code review/discuss this part

Reply:
This is obsolete as the state check was refactored following another approach. Marking as resolved.